### PR TITLE
vim/shared.lua: Add vim.tbl_find method

### PR DESCRIPTION
**Pull request by h-michael**
discuss at https://github.com/neovim/neovim/pull/11603#issuecomment-569483266

### CODE REVIEW

**Comment by h-michael** [comment]
@justinmk gentle ping

File: runtime/lua/vim/shared.lua [hunk 1]
```diff
@@ -300,6 +300,29 @@ function vim.tbl_islist(t)
...
+  end
+
+  return function(t, o)
+    vim.validate{t={t, 't'}}
+    for _, v in pairs(t) do
+      if _apply(v, o) then return v end
+    end
+    return nil
+  end
+end)()
```
Comment (justinmk) [inline runtime/lua/vim/shared.lua]: I think we don't need this pattern here. And we can avoid `_apply`  checking  if o is callable for each comparison. Something like this (untested):

```
function vim.tbl_find(t, o)
  vim.validate{t={t, 't'}}
  local cmp = vim.is_callable(o) and o or function(v) o == v end
  for _, v in pairs(t) do
    if cmp(v) then return v end
  end
  return nil
end
```

File: runtime/lua/vim/shared.lua [hunk 2]
```diff
@@ -300,6 +300,29 @@ function vim.tbl_islist(t)
...
 end
 
+--- Returns the first element from the table that matches the second argument.
+---
+--@param t table
+--@param o Any object. If f is callable, apply the element to f, otherwise compare the element to f.
+function vim.tbl_find(t, o) end -- luacheck: no unused
+vim.tbl_find = (function()
+  local function _apply(v, o)
+    if vim.is_callable(o) then
```
Comment (justinmk) [inline runtime/lua/vim/shared.lua]: @blueyed coverage report claims this line isn't hit?

**Comment by h-michael** [comment]
I've fixed. But I don't know why c codes coverage decrease.

**Comment by justinmk** [comment]
[lume.find()](https://github.com/rxi/lume#lumefindt-value) returns the key instead of the index, if the table arg is dict-like. 

https://github.com/rxi/lume/blob/d8c2eddc10af994ad4956cf0b7ae7188e86db47e/lume.lua#L57

Any opinions on whether we want to return key for dict-like? cc @bfredl @norcalli

**Comment by norcalli** [comment]
I do something like this in my find definitions:

```lua
function vim.find(t, f)
  for k, v in pairs(t) do
    local r = f(v, k)
    if r then
      return r
    end
  end
end
```

**Comment by norcalli** [comment]
PS If we ever want to do a stream/iterable like library, then we should possibly consider that in the naming of the function. Although we could put that all under a `vim.iter.` kind of thing.

**Comment by h-michael** [comment]
> PS If we ever want to do a stream/iterable like library, then we should possibly consider that in the naming of the function. Although we could put that all under a vim.iter. kind of thing.

I agree with it. But that will be discussed in another issue/pr.

**Comment by h-michael** [comment]
@justinmk 
1. return value even if the table is a list.
2. default return value, but can return key with option
3. return key(index) like rume.find
4. @norcalli suggested

**Comment by lewis6991** [comment]
We now have `tbl_contains`, `tbl_filter` and `tbl_map`.

**Pull request closed without merging**